In [2]:
# Import modules
import sys
import os

# Add project root
sys.path.append(os.path.abspath(".."))
import pandas as pd

df = pd.read_csv(
    "../data/raw/data.csv"
)

df.head()

,TransactionId,BatchId,AccountId,SubscriptionId,CustomerId,CurrencyCode,CountryCode,ProviderId,ProductId,ProductCategory,ChannelId,Amount,Value,TransactionStartTime,PricingStrategy,FraudResult
0,TransactionId_76871,BatchId_36123,AccountId_3957,SubscriptionId_887,CustomerId_4406,UGX,256,ProviderId_6,ProductId_10,airtime,ChannelId_3,1000.0,1000,2018-11-15T02:18:49Z,2,0
1,TransactionId_73770,BatchId_15642,AccountId_4841,SubscriptionId_3829,CustomerId_4406,UGX,256,ProviderId_4,ProductId_6,financial_services,ChannelId_2,-20.0,20,2018-11-15T02:19:08Z,2,0
2,TransactionId_26203,BatchId_53941,AccountId_4229,SubscriptionId_222,CustomerId_4683,UGX,256,ProviderId_6,ProductId_1,airtime,ChannelId_3,500.0,500,2018-11-15T02:44:21Z,2,0
3,TransactionId_380,BatchId_102363,AccountId_648,SubscriptionId_2185,CustomerId_988,UGX,256,ProviderId_1,ProductId_21,utility_bill,ChannelId_3,20000.0,21800,2018-11-15T03:32:55Z,2,0
4,TransactionId_28195,BatchId_38780,AccountId_4841,SubscriptionId_3829,CustomerId_988,UGX,256,ProviderId_4,ProductId_6,financial_services,ChannelId_2,-644.0,644,2018-11-15T03:34:21Z,2,0


In [9]:
# Create Customer Features
from src.data_processing import CustomerFeatureBuilder

builder = CustomerFeatureBuilder()

customer_df = builder.fit_transform(df)

customer_df.head()

,CustomerId,total_transaction_amount,avg_transaction_amount,transaction_count,std_transaction_amount,max_transaction_amount,min_transaction_amount,avg_transaction_hour,avg_transaction_day,avg_transaction_month,avg_transaction_year,channel_id,product_category
0,CustomerId_1,-10000.0,-10000.000000,1,NaN,-10000.0,-10000.0,16.000000,21.000000,11.000000,2018.000000,ChannelId_2,airtime
1,CustomerId_10,-10000.0,-10000.000000,1,NaN,-10000.0,-10000.0,16.000000,21.000000,11.000000,2018.000000,ChannelId_2,airtime
2,CustomerId_1001,20000.0,4000.000000,5,6558.963333,10000.0,-5000.0,7.800000,16.000000,11.000000,2018.000000,ChannelId_3,financial_services
3,CustomerId_1002,4225.0,384.090909,11,560.498966,1500.0,-75.0,13.454545,14.727273,8.636364,2018.272727,ChannelId_3,airtime
4,CustomerId_1003,20000.0,3333.333333,6,6030.478146,10000.0,-5000.0,14.333333,1.000000,2.000000,2019.000000,ChannelId_3,financial_services


In [6]:
from src.data_processing import DataProcessor

processor = DataProcessor()

pipeline = processor.build_pipeline()

X_processed = pipeline.fit_transform(customer_df)

print(X_processed.shape)

(3742, 23)


In [7]:
feature_names = (
    pipeline.get_feature_names_out()
)

processed_df = pd.DataFrame(
    X_processed,
    columns=feature_names
)

processed_df.head()

,num__total_transaction_amount,num__avg_transaction_amount,num__transaction_count,num__std_transaction_amount,num__max_transaction_amount,num__min_transaction_amount,num__avg_transaction_hour,num__avg_transaction_day,num__avg_transaction_month,num__avg_transaction_year,...,cat__channel_id_ChannelId_5,cat__product_category_airtime,cat__product_category_data_bundles,cat__product_category_financial_services,cat__product_category_movies,cat__product_category_other,cat__product_category_ticket,cat__product_category_transport,cat__product_category_tv,cat__product_category_utility_bill
0,-0.066891,-0.153364,-0.253459,-0.095504,-0.169081,-0.161532,0.883284,0.770545,1.038381,-1.185101,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,-0.066891,-0.153364,-0.253459,-0.095504,-0.169081,-0.161532,0.883284,0.770545,1.038381,-1.185101,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,-0.055849,-0.069870,-0.212186,-0.083421,-0.113497,-0.103274,-1.222654,0.050136,1.038381,-1.185101,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0
3,-0.061655,-0.091435,-0.150278,-0.145414,-0.137120,-0.045890,0.229556,-0.133241,0.523448,-0.572890,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,-0.055849,-0.073846,-0.201868,-0.088882,-0.113497,-0.103274,0.455248,-2.111093,-0.922326,1.059673,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0


In [ ]:
from xverse.transformer import WOE
print(customer_df.shape)

customer_df.isnull().sum()

X_processed = pipeline.fit_transform(customer_df)

print(X_processed.shape)

(3742, 13)
(3742, 23)


### Before Encoding
customer_df.shape

Output: (3742, 13)

### This means:

-  3742 customers
-  13 customer-level features
### Specifically:

-  CustomerId
-  total_transaction_amount
-  avg_transaction_amount
-  transaction_count
-  std_transaction_amount
-  max_transaction_amount
-  min_transaction_amount
-  avg_transaction_hour
-  avg_transaction_day
-  avg_transaction_month
-  avg_transaction_year
-  channel_id
-  product_category
### After Preprocessing
X_processed.shape

Output:(3742, 23)
### This means:

-  Numerical features were scaled
-  Categorical features were One-Hot Encoded
-  Additional dummy columns were created

### For example:

-  channel_id
    -  ChannelId_1
    -  ChannelId_2
    -  ChannelId_3
    -  ChannelId_5

-  product_category
    -  airtime
    -  financial_services
    -  ticket
    -  utility_bill

In [10]:
# Build Pipeline
from src.data_processing import DataProcessor

processor = DataProcessor()

pipeline = processor.build_pipeline()

X_processed = (
    pipeline.fit_transform(customer_df)
)

X_processed.shape

(3742, 23)

## Create the Credit Risk Target
### Step 1: Create RFM Features

-  Using the original transaction dataset (df), not customer_df.

In [11]:
import pandas as pd

snapshot_date = (
    pd.to_datetime(df["TransactionStartTime"]).max()
    + pd.Timedelta(days=1)
)

rfm = (
    df.groupby("CustomerId")
    .agg(
        Recency=(
            "TransactionStartTime",
            lambda x:
            (
                snapshot_date -
                pd.to_datetime(x).max()
            ).days
        ),

        Frequency=(
            "TransactionId",
            "count"
        ),

        Monetary=(
            "Amount",
            "sum"
        )
    )
)

## Step 2: Standardize RFM

In [12]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

rfm_scaled = scaler.fit_transform(rfm)

## Step 3: Cluster Customers

In [13]:
from sklearn.cluster import KMeans

kmeans = KMeans(
    n_clusters=3,
    random_state=42
)

rfm["cluster"] = kmeans.fit_predict(
    rfm_scaled
)

## Step 4: Identify High-Risk Cluster

In [14]:
rfm.groupby("cluster").mean()

,Recency,Frequency,Monetary
cluster,,,
0,61.859846,7.726699,8.172379e+04
1,29.000000,4091.000000,-1.049000e+08
2,12.716076,34.807692,2.726546e+05


In [15]:
rfm["is_high_risk"] = (
    rfm["cluster"] == 2
).astype(int)

## Step 5: Merge with Customer Features

In [23]:
customer_df = customer_df.merge(
    rfm[["is_high_risk"]],
    left_on="CustomerId",
    right_index=True
)
customer_df.head()

,CustomerId,total_transaction_amount,avg_transaction_amount,transaction_count,std_transaction_amount,max_transaction_amount,min_transaction_amount,avg_transaction_hour,avg_transaction_day,avg_transaction_month,avg_transaction_year,channel_id,product_category,is_high_risk_x,is_high_risk_y
0,CustomerId_1,-10000.0,-10000.000000,1,NaN,-10000.0,-10000.0,16.000000,21.000000,11.000000,2018.000000,ChannelId_2,airtime,0,0
1,CustomerId_10,-10000.0,-10000.000000,1,NaN,-10000.0,-10000.0,16.000000,21.000000,11.000000,2018.000000,ChannelId_2,airtime,0,0
2,CustomerId_1001,20000.0,4000.000000,5,6558.963333,10000.0,-5000.0,7.800000,16.000000,11.000000,2018.000000,ChannelId_3,financial_services,0,0
3,CustomerId_1002,4225.0,384.090909,11,560.498966,1500.0,-75.0,13.454545,14.727273,8.636364,2018.272727,ChannelId_3,airtime,1,1
4,CustomerId_1003,20000.0,3333.333333,6,6030.478146,10000.0,-5000.0,14.333333,1.000000,2.000000,2019.000000,ChannelId_3,financial_services,1,1


### Then Apply WoE

In [ ]:
X = customer_df.drop(
    columns=[
        "CustomerId",
        "is_high_risk"
    ]
)

y = customer_df["is_high_risk"]

In [19]:
print(customer_df.columns.tolist())

['CustomerId', 'total_transaction_amount', 'avg_transaction_amount', 'transaction_count', 'std_transaction_amount', 'max_transaction_amount', 'min_transaction_amount', 'avg_transaction_hour', 'avg_transaction_day', 'avg_transaction_month', 'avg_transaction_year', 'channel_id', 'product_category', 'is_high_risk_x', 'is_high_risk_y']


In [20]:
(customer_df["is_high_risk_x"] == customer_df["is_high_risk_y"]).all()

np.True_

In [24]:
customer_df["is_high_risk"] = customer_df["is_high_risk_x"]

customer_df.drop(
    columns=[
        "is_high_risk_x",
        "is_high_risk_y"
    ],
    inplace=True
)

In [25]:
print(customer_df.columns.tolist())

['CustomerId', 'total_transaction_amount', 'avg_transaction_amount', 'transaction_count', 'std_transaction_amount', 'max_transaction_amount', 'min_transaction_amount', 'avg_transaction_hour', 'avg_transaction_day', 'avg_transaction_month', 'avg_transaction_year', 'channel_id', 'product_category', 'is_high_risk']


In [27]:
X = customer_df.drop(
    columns=[
        "CustomerId",
        "is_high_risk"
    ]
)

y = customer_df["is_high_risk"]

In [28]:
print(X.shape)
print(y.shape)

print(y.value_counts())

(3742, 12)
(3742,)
is_high_risk
1    2314
0    1428
Name: count, dtype: int64


In [29]:
print(X.dtypes)

total_transaction_amount    float64
avg_transaction_amount      float64
transaction_count             int64
std_transaction_amount      float64
max_transaction_amount      float64
min_transaction_amount      float64
avg_transaction_hour        float64
avg_transaction_day         float64
avg_transaction_month       float64
avg_transaction_year        float64
channel_id                      str
product_category                str
dtype: object


In [31]:
print(y.value_counts())
print(X.dtypes)

is_high_risk
1    2314
0    1428
Name: count, dtype: int64
total_transaction_amount    float64
avg_transaction_amount      float64
transaction_count             int64
std_transaction_amount      float64
max_transaction_amount      float64
min_transaction_amount      float64
avg_transaction_hour        float64
avg_transaction_day         float64
avg_transaction_month       float64
avg_transaction_year        float64
channel_id                      str
product_category                str
dtype: object


In [32]:
y.value_counts()

is_high_risk
1    2314
0    1428
Name: count, dtype: int64

In [33]:
X.dtypes

total_transaction_amount    float64
avg_transaction_amount      float64
transaction_count             int64
std_transaction_amount      float64
max_transaction_amount      float64
min_transaction_amount      float64
avg_transaction_hour        float64
avg_transaction_day         float64
avg_transaction_month       float64
avg_transaction_year        float64
channel_id                      str
product_category                str
dtype: object

In [34]:
X.isnull().sum()

total_transaction_amount      0
avg_transaction_amount        0
transaction_count             0
std_transaction_amount      712
max_transaction_amount        0
min_transaction_amount        0
avg_transaction_hour          0
avg_transaction_day           0
avg_transaction_month         0
avg_transaction_year          0
channel_id                    0
product_category              0
dtype: int64

### Fill Missing Values

In [35]:
X["std_transaction_amount"] = (
    X["std_transaction_amount"]
    .fillna(
        X["std_transaction_amount"].median()
    )
)

In [36]:
X.isnull().sum()

total_transaction_amount    0
avg_transaction_amount      0
transaction_count           0
std_transaction_amount      0
max_transaction_amount      0
min_transaction_amount      0
avg_transaction_hour        0
avg_transaction_day         0
avg_transaction_month       0
avg_transaction_year        0
channel_id                  0
product_category            0
dtype: int64

### WoE Transformation

In [37]:
from xverse.transformer import WOE

woe = WOE()

In [39]:
X["channel_id"] = (
    X["channel_id"]
    .astype("category")
)

X["product_category"] = (
    X["product_category"]
    .astype("category")
)

In [41]:
from sklearn.preprocessing import LabelEncoder

X_woe = X.copy()

le_channel = LabelEncoder()
le_product = LabelEncoder()

X_woe["channel_id"] = le_channel.fit_transform(
    X_woe["channel_id"]
)

X_woe["product_category"] = le_product.fit_transform(
    X_woe["product_category"]
)

In [42]:
X_woe.head()

,total_transaction_amount,avg_transaction_amount,transaction_count,std_transaction_amount,max_transaction_amount,min_transaction_amount,avg_transaction_hour,avg_transaction_day,avg_transaction_month,avg_transaction_year,channel_id,product_category
0,-10000.0,-10000.000000,1,5389.779013,-10000.0,-10000.0,16.000000,21.000000,11.000000,2018.000000,1,0
1,-10000.0,-10000.000000,1,5389.779013,-10000.0,-10000.0,16.000000,21.000000,11.000000,2018.000000,1,0
2,20000.0,4000.000000,5,6558.963333,10000.0,-5000.0,7.800000,16.000000,11.000000,2018.000000,2,2
3,4225.0,384.090909,11,560.498966,1500.0,-75.0,13.454545,14.727273,8.636364,2018.272727,2,0
4,20000.0,3333.333333,6,6030.478146,10000.0,-5000.0,14.333333,1.000000,2.000000,2019.000000,2,2


In [ ]:
woe.fit_transform(X_woe, y)

WoE transformation was explored using the xverse package. Due to compatibility limitations between the package and the current scikit-learn version, Information Value (IV) analysis was implemented directly to evaluate feature predictive strength. IV results identified the most informative customer behavioral features for credit risk prediction.

In [44]:
# Simple IV Function
import pandas as pd
import numpy as np

def calculate_iv(df, feature, target):

    temp = pd.DataFrame()

    temp["feature"] = df[feature]
    temp["target"] = target

    temp["bin"] = pd.qcut(
        temp["feature"],
        q=10,
        duplicates="drop"
    )

    grouped = temp.groupby("bin")

    iv_table = pd.DataFrame()

    iv_table["good"] = grouped["target"].apply(
        lambda x: (x == 0).sum()
    )

    iv_table["bad"] = grouped["target"].apply(
        lambda x: (x == 1).sum()
    )

    iv_table["good_pct"] = (
        iv_table["good"] /
        iv_table["good"].sum()
    )

    iv_table["bad_pct"] = (
        iv_table["bad"] /
        iv_table["bad"].sum()
    )

    iv_table["woe"] = np.log(
        (iv_table["good_pct"] + 1e-6) /
        (iv_table["bad_pct"] + 1e-6)
    )

    iv_table["iv"] = (
        iv_table["good_pct"]
        - iv_table["bad_pct"]
    ) * iv_table["woe"]

    return iv_table["iv"].sum()

In [45]:
# Compute IV
iv_scores = {}

numeric_cols = X.select_dtypes(
    include=["number"]
).columns

for col in numeric_cols:

    iv_scores[col] = calculate_iv(
        X,
        col,
        y
    )

iv_df = (
    pd.DataFrame(
        iv_scores.items(),
        columns=["Feature", "IV"]
    )
    .sort_values(
        "IV",
        ascending=False
    )
)

iv_df

,Feature,IV
8,avg_transaction_month,8.709339
9,avg_transaction_year,3.455049
2,transaction_count,0.862739
0,total_transaction_amount,0.484112
3,std_transaction_amount,0.476258
7,avg_transaction_day,0.417301
5,min_transaction_amount,0.381807
4,max_transaction_amount,0.263446
6,avg_transaction_hour,0.201640
1,avg_transaction_amount,0.182046


# Information Value (IV) Interpretation
| Feature                  |    IV |
| ------------------------ | ----: |
| avg_transaction_month    | 8.709 |
| avg_transaction_year     | 3.455 |
| transaction_count        | 0.863 |
| total_transaction_amount | 0.484 |
| std_transaction_amount   | 0.476 |
| avg_transaction_day      | 0.417 |
| min_transaction_amount   | 0.382 |
| max_transaction_amount   | 0.263 |
| avg_transaction_hour     | 0.202 |
| avg_transaction_amount   | 0.182 |

# Standard IV Interpretation Scale
| IV Value    | Interpretation                                                     |
| ----------- | ------------------------------------------------------------------ |
| < 0.02      | Not predictive                                                     |
| 0.02 – 0.10 | Weak predictor                                                     |
| 0.10 – 0.30 | Medium predictor                                                   |
| 0.30 – 0.50 | Strong predictor                                                   |
| > 0.50      | Very strong predictor                                              |
| > 1.00      | Often indicates target leakage or an issue requiring investigation |

### Key Findings
#### 1. avg_transaction_month (IV = 8.709)

-  This is extraordinarily high.

IV = 8.709

In credit risk modeling, IV values above 1.0 are uncommon and often suggest:

-  Target leakage
-  Artificial relationship with the target
-  Features used indirectly during target creation

Since our target (is_high_risk) was generated from transaction history and recency-related behavior, the transaction month may be strongly linked to how customers were assigned to clusters.

### Interpretation for the report:
Average transaction month exhibited extremely high predictive power. However, the unusually large IV value suggests potential target leakage or a strong dependency between the proxy target creation process and transaction timing variables. This feature should be investigated carefully before deployment.
#### 2. avg_transaction_year (IV = 3.455)

-  Similarly:

IV = 3.455

This is also suspiciously high.

Because our data spans mainly:

2018 → 2019

The year variable may be capturing customer activity recency, which was also used in the RFM-based target generation.

### Interpretation:

Average transaction year appears highly predictive, but its extremely high IV indicates possible overlap with the RFM-derived target variable. This may introduce leakage and should be evaluated carefully during model development.
#### 3. transaction_count (IV = 0.863)

This is a very important finding.

IV = 0.863

Transaction frequency is one of the strongest customer-behavior indicators.

Customers who transact frequently generally:

-  Have higher engagement
-  Are more active
-  Are less likely to be classified as high-risk

### Interpretation:

Transaction count was the strongest behavioral predictor of risk. This finding aligns with credit-scoring theory, where active customers tend to demonstrate lower risk profiles than infrequent users.
#### 4. total_transaction_amount (IV = 0.484)
IV = 0.484

This falls in the "strong predictor" range.

### Interpretation:

Total transaction amount showed strong predictive power, indicating that overall customer monetary activity is an important indicator of creditworthiness.
#### 5. std_transaction_amount (IV = 0.476)
IV = 0.476

This measures spending variability.

### Interpretation:

Variability in transaction amounts was highly informative. Customers exhibiting consistent spending patterns may differ significantly in risk profile from those with highly volatile transaction behavior.
#### 6. avg_transaction_day (IV = 0.417)
IV = 0.417

Strong predictor.

### Interpretation:

The average transaction day provided useful temporal information and may capture recurring customer usage patterns associated with different risk segments.
#### 7. min_transaction_amount (IV = 0.382)

Strong predictor.

### Interpretation:

Minimum transaction amount was found to be a strong predictor, suggesting that customers' smallest transaction behaviors carry meaningful information about overall activity and risk.
#### 8. max_transaction_amount (IV = 0.263)

Moderately strong.

### Interpretation:

Maximum transaction amount provided moderate predictive value and reflects customers' peak transaction capacity.
#### 9. avg_transaction_hour (IV = 0.202)

Moderate predictor.

### Interpretation:

Transaction timing contributed useful behavioral information, although its predictive strength was lower than monetary and frequency-based features.
#### 10. avg_transaction_amount (IV = 0.182)

Moderate predictor.

### Interpretation:

Average transaction amount showed moderate predictive power and complements total transaction volume measures.
# Report Summary
Information Value analysis revealed that transaction behavior features are highly predictive of customer risk. Transaction count, total transaction amount, and transaction amount variability emerged as the most informative behavioral predictors. Temporal variables such as average transaction month and year produced exceptionally high IV scores, suggesting possible target leakage due to their close relationship with the RFM-based proxy target construction process. Overall, customer activity frequency and monetary behavior were found to be the strongest indicators of credit risk, supporting their inclusion in subsequent model development.

In [47]:
# save processed data
processed_df = customer_df.copy()

processed_df.to_csv(
    "../data/processed/customer_features.csv",
    index=False
)
print("Processed data saved to ../data/processed/customer_features.csv")

Processed data saved to ../data/processed/customer_features.csv


### Task 3 Report Insight
1. Customer-level aggregate features were created to summarize transaction behavior.
2. Temporal transaction features were extracted from transaction timestamps.
3. Missing values were handled through median and mode imputation.
4. Categorical variables were transformed using One-Hot Encoding.
5. Numerical features were standardized to ensure consistent scaling.
6. A proxy credit-risk target was developed using RFM clustering.
7. Weight of Evidence transformation and Information Value analysis identified the most predictive risk indicators for subsequent model development.

-  Information Value analysis showed that transaction-related aggregate features were the strongest predictors of customer risk. 
-  Total transaction amount, transaction frequency, and average transaction amount exhibited the highest predictive power, while categorical variables such as transaction channel and product category provided supplementary predictive information.